### PR TITLE
CLOUDP-57841: Delete dbusers C/OM

### PR DESCRIPTION
**Pull request by gssbzn**
## Proposed changes

_Jira ticket:_ [CLOUDP-57841](https://jira.mongodb.org/browse/CLOUDP-57841)

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run `make fmt` and formatted my code

### CODE REVIEW

File: internal/cli/atlas_dbusers_create.go [hunk 1]
```diff
@@ -58,7 +59,7 @@ func (opts *atlasDBUsersCreateOpts) Run() error {
 
 func (opts *atlasDBUsersCreateOpts) newDatabaseUser() *atlas.DatabaseUser {
 	return &atlas.DatabaseUser{
-		DatabaseName: convert.AdminDB,
+		DatabaseName: opts.authDB,
```
Comment (gssbzn) [inline internal/cli/atlas_dbusers_create.go]: we missed this in the last PR

File: internal/cli/cli.go [hunk 2]
```diff
@@ -24,7 +25,7 @@ import (
 
 const (
 	fallbackSuccessMessage = "'%s' deleted\n"
-	fallbackFailMessage    = "'%s' not deleted\n"
+	fallbackFailMessage    = "entry not deleted"
```
Comment (gssbzn) [inline internal/cli/cli.go]: other fail messages don't interpolate so making the fall back compatible

File: internal/cli/cli.go [hunk 3]
```diff
@@ -67,83 +68,32 @@ type deleteOpts struct {
...
-type DeleterFromProject func(projectID string, entry string) error
-
-// DeleterFromProjectAuthDB a function to delete from the store.
-type DeleterFromProjectAuthDB func(authDB string, projectID string, entry string) error
-
-// DeleteFromProject deletes a resource from a project, it expects a callback
+// Delete deletes a resource not associated to a project, it expects a callback
 // that should perform the deletion from the store.
-func (opts *deleteOpts) DeleteFromProject(d DeleterFromProject, projectID string) error {
+func (opts *deleteOpts) Delete(d interface{}, a ...string) error {
```
Comment (gssbzn) [inline internal/cli/cli.go]: this function can do now all three functions

File: internal/cli/cli.go [hunk 4]
```diff
@@ -67,83 +68,32 @@ type deleteOpts struct {
...
-// DeleterFromProjectAuthDB deletes a resource from a project, it expects a callback
-// that should perform the deletion from the store.
-func (opts *deleteOpts) DeleterFromProjectAuthDB(d DeleterFromProjectAuthDB, authDB, projectID string) error {
-	if !opts.confirm {
-		opts.printFailMessage()
-		return nil
+	var err error
+	fmt.Printf("%#v", d)
+	switch f := d.(type) {
+	case func(string) error:
```
Comment (gssbzn) [inline internal/cli/cli.go]: for some reason I could not use types here so getting rid of them

File: internal/cli/cli.go [hunk 5]
```diff
@@ -157,3 +107,19 @@ func (opts *deleteOpts) Confirm() error {
...
+	return fallbackSuccessMessage
+}
+
+// FailMessage gets the set fail message or the default value
+func (opts *deleteOpts) FailMessage() string {
+	if opts.failMessage != "" {
+		return opts.failMessage
+	}
+	return fallbackFailMessage
+}
```
Comment (gssbzn) [inline internal/cli/cli.go]: making this just getters and leaving the printing out

File: internal/cli/cli.go [hunk 6]
```diff
@@ -67,83 +68,32 @@ type deleteOpts struct {
...
-}
 
-// DeleterFromProjectAuthDB deletes a resource from a project, it expects a callback
-// that should perform the deletion from the store.
-func (opts *deleteOpts) DeleterFromProjectAuthDB(d DeleterFromProjectAuthDB, authDB, projectID string) error {
-	if !opts.confirm {
-		opts.printFailMessage()
-		return nil
+	var err error
+	fmt.Printf("%#v", d)
```
Comment (andreaangiolillo) [inline internal/cli/cli.go]: [qa] why do we have a print here?

File: internal/convert/database_user_test.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
+package convert
```
Comment (gssbzn) [inline internal/convert/database_user_test.go]: we were missing tests for this

File: internal/cli/cli.go [hunk 8]
```diff
@@ -67,83 +68,32 @@ type deleteOpts struct {
...
+	switch f := d.(type) {
```
Comment (andreaangiolillo) [inline internal/cli/cli.go]: wow! 👏 😮

File: internal/cli/cli.go (hunk 6 above)
Comment (gssbzn) [reply internal/cli/cli.go]: ha, good catch, I think this is a left over from em doing some testing on why typing wouldn't work

File: internal/cli/ops_manager_dbusers_delete.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+	cmd := &cobra.Command{
+		Use:     "delete [username]",
+		Short:   "Delete a database user for a project.",
+		Aliases: []string{"rm"},
+		Args:    cobra.ExactArgs(1),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			if err := opts.init(); err != nil {
+				return err
+			}
+			opts.entry = args[0]
```
Comment (andreaangiolillo) [inline internal/cli/ops_manager_dbusers_delete.go]: [qa] Do you think it is better to set the arg inside PreRun or Run?. I am quite sure that we are setting that arg inside Run in other commands. We should decide one place and change all the other commands in accordance with that decision. I honestly think that the preRun would be a better place 👍

File: internal/cli/ops_manager_dbusers_delete.go [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return opts.Run()
+		},
+	}
+
+	cmd.Flags().StringVar(&opts.authDB, flags.AuthDB, convert.AdminDB, usage.AuthDB)
+
+	cmd.Flags().StringVar(&opts.projectID, flags.ProjectID, "", usage.ProjectID)
+
+	_ = cmd.MarkFlagRequired(flags.Username)
```
Comment (andreaangiolillo) [inline internal/cli/ops_manager_dbusers_delete.go]: [qa] why are we using this? username is not a flag right?

Comment (gssbzn) [reply internal/cli/ops_manager_dbusers_delete.go]: oops copy/paste leftover

File: internal/cli/ops_manager_dbusers_delete.go (hunk 9 above)
Comment (gssbzn) [reply internal/cli/ops_manager_dbusers_delete.go]: Yeah we definitely need to agree on this, I've moved it around based on need, the prompt makes sense to run on the pre run so we need the entry set before prompting, but some validations don't run before preRun, I know flags don't do it but I can't remember if the number of args validation does, if we can confirm it does then we should move them all to a preRun

**Comment by gssbzn** [comment]
@andreaangiolillo RFAL

**Pull request merged**